backend/src/main/java/ai/verta/modeldb/cron_jobs/DeleteEntitiesCron.java

```diff
@@ -0,0 +1,666 @@
+              .append("SET dv.")
+              .append(ModelDBConstants.DELETED)
+              .append(" = :deleted ")
+              .append(" WHERE dv.")
+              .append(ModelDBConstants.DATASET_ID)
+              .append(" IN (:datasetIds)")
+              .toString();
+      Query deletedDatasetVersionQuery =
+          session.createQuery(updateDeletedStatusDatasetVersionQueryString);
+      deletedDatasetVersionQuery.setParameter("deleted", true);
+      deletedDatasetVersionQuery.setParameter("datasetIds", datasetIds);
+      deletedDatasetVersionQuery.executeUpdate();
+
+      for (DatasetEntity datasetEntity : datasetEntities) {
+        session.delete(datasetEntity);
+      }
+      transaction.commit();
+    }
+    LOGGER.debug("Dataset Deleted successfully : Deleted datasets count {}", datasetIds.isEmpty());
+  }
+
+  private void deleteRoleBindingsOfAccessibleDatasets(
+      List<DatasetEntity> allowedDatasets, List<String> roleBindingNames) {
+    UserInfo unsignedUser = authService.getUnsignedUser();
+    for (DatasetEntity datasetEntity : allowedDatasets) {
+      String datasetId = datasetEntity.getId();
+
+      String ownerRoleBindingName =
+          roleService.buildRoleBindingName(
+              ModelDBConstants.ROLE_DATASET_OWNER,
+              datasetEntity.getId(),
+              datasetEntity.getOwner(),
+              ModelResourceEnum.ModelDBServiceResourceTypes.DATASET.name());
+      if (ownerRoleBindingName != null) {
+        roleBindingNames.add(ownerRoleBindingName);
+      }
+
+      if (datasetEntity.getDataset_visibility()
+          == DatasetVisibilityEnum.DatasetVisibility.PUBLIC.getNumber()) {
+        String publicReadRoleBindingName =
+            roleService.buildRoleBindingName(
+                ModelDBConstants.ROLE_DATASET_PUBLIC_READ,
+                datasetId,
+                authService.getVertaIdFromUserInfo(unsignedUser),
+                ModelResourceEnum.ModelDBServiceResourceTypes.DATASET.name());
+        if (publicReadRoleBindingName != null && !publicReadRoleBindingName.isEmpty()) {
+          roleBindingNames.add(publicReadRoleBindingName);
+        }
+      }
+
+      // Delete workspace based roleBindings
+      List<String> workspaceRoleBindingNames =
+          getWorkspaceRoleBindingsForDataset(
+              datasetEntity.getWorkspace(),
+              WorkspaceTypeEnum.WorkspaceType.forNumber(datasetEntity.getWorkspace_type()),
+              datasetEntity.getId(),
+              DatasetVisibilityEnum.DatasetVisibility.forNumber(
+                  datasetEntity.getDataset_visibility()));
+      if (!workspaceRoleBindingNames.isEmpty()) {
+        roleBindingNames.addAll(workspaceRoleBindingNames);
+      }
+    }
+
+    // Remove all datasetEntity collaborators
+    roleService.deleteAllResources(
+        allowedDatasets.stream().map(DatasetEntity::getId).collect(Collectors.toList()),
+        ModelResourceEnum.ModelDBServiceResourceTypes.DATASET);
```

Comment:
Same.

Reply:
Created separate functions called `deleteRoleBindingsFor*` and moved all delete role bindings code in that function for each entity.